llarp/nodedb.cpp

```diff
@@ -99,6 +99,73 @@ namespace llarp
     return m_Root / skiplistDir / fname;
   }
 
+  void
+  NodeDB::set_bootstrap_routers(const std::set<RemoteRC>& rcs)
+  {
+    bootstraps.clear();  // this function really shouldn't be called more than once, but...
+    for (const auto& rc : rcs)
+      bootstraps.emplace(rc.router_id(), rc);
+  }
+
+  void
+  NodeDB::set_router_whitelist(
+      const std::vector<RouterID>& whitelist,
+      const std::vector<RouterID>& greylist,
+      const std::vector<RouterID>& greenlist)
+  {
+    if (whitelist.empty())
+      return;
+
+    registered_routers.clear();
+    registered_routers.insert(whitelist.begin(), whitelist.end());
+    registered_routers.insert(greylist.begin(), greylist.end());
+    registered_routers.insert(greenlist.begin(), greenlist.end());
+
+    router_whitelist.clear();
+    router_whitelist.insert(whitelist.begin(), whitelist.end());
+    router_greylist.clear();
+    router_greylist.insert(greylist.begin(), greylist.end());
+    router_greenlist.clear();
+    router_greenlist.insert(greenlist.begin(), greenlist.end());
+
+    log::info(
+        logcat, "lokinet service node list now has ", router_whitelist.size(), " active routers");
+  }
+
+  std::optional<RouterID>
+  NodeDB::get_random_whitelist_router() const
+  {
+    const auto sz = router_whitelist.size();
+    if (sz == 0)
+      return std::nullopt;
+    auto itr = router_whitelist.begin();
+    if (sz > 1)
+      std::advance(itr, randint() % sz);
+    return *itr;
+  }
+
+  bool
+  NodeDB::is_connection_allowed(const RouterID& remote) const
+  {
+    if (pinned_edges.size() && pinned_edges.count(remote) == 0 && !bootstraps.count(remote))
```

Comment:
Hilariously minor:
So minor that I'm laughing writing it. You alternate between `container.count(target) == 0` and `!container.count(target)` and periodically a `not containter.count(target)` (maybe that third one is a residual formalism from my terrible practice, since I usually put `not` instead of `!`)

Reply:
Sometimes inconsistencies like that stick out to me like a sore thumb and other times they're just background noise...

I agree this should be consistent, but I always waffle back and forth on how to express "`count()` is or is not zero" (bool or integer compare).  Will address in future commit in PR (or yell at me if I forget).